src/normalize/normalize.rs

```diff
@@ -0,0 +1,149 @@
+use std::io::Result;
+use std::path::PathBuf;
+
+use crate::elf;
+use crate::elf::ElfParser;
+use crate::log;
+use crate::maps;
+use crate::maps::MapsEntry;
+use crate::maps::Pid;
+use crate::util;
+use crate::Addr;
+
+use super::meta::Binary;
+use super::meta::UserAddrMeta;
+
+
+/// A type capturing normalized addresses along with captured meta data.
+///
+/// This type enables "remote" symbolization. That is to say, it represents the
+/// input necessary for addresses to be symbolized on a system other than where
+/// they were recorded.
+#[derive(Clone, Debug)]
+pub struct NormalizedAddrs<M> {
+    /// Normalized addresses along with an index into `meta` for retrieval of
+    /// the corresponding [`AddrMeta`] information.
+    ///
+    /// A normalized address is one as it would appear in a binary or debug
+    /// symbol file, i.e., one excluding any relocations.
+    pub addrs: Vec<(Addr, usize)>,
+    /// Meta information about the normalized addresses.
+    pub meta: Vec<M>,
+}
+
+/// A type representing normalized user addresses.
+pub type NormalizedUserAddrs = NormalizedAddrs<UserAddrMeta>;
+
+
+/// Normalize a virtual address belonging to an ELF file represented by the
+/// provided [`MapsEntry`].
+fn normalize_elf_addr(virt_addr: Addr, entry: &MapsEntry) -> Result<Addr> {
+    let parser = ElfParser::open(&entry.path)?;
+    let phdrs = parser.program_headers()?;
+    let addr = phdrs
+        .iter()
+        .find_map(|phdr| {
+            if phdr.p_type == elf::types::PT_LOAD {
+                if (phdr.p_offset..phdr.p_offset + phdr.p_memsz).contains(&entry.offset) {
+                    let segment_offset = phdr.p_offset - entry.offset;
+                    return Some(virt_addr - entry.range.start - segment_offset as Addr)
+                }
+            }
+            None
+        })
+        .ok_or_else(|| {
+            Error::new(
+                ErrorKind::InvalidInput,
+                format!(
+                    "failed to find ELF segment in {} that contains file offset 0x{:x}",
+                    entry.path.display(),
+                    entry.offset,
+                ),
+            )
+        })?;
+
+    Ok(addr)
+}
+
+
+/// Normalize `addresses` belonging to a process.
+///
+/// Normalize all `addresses` in a given process. For optimum performance, it is
+/// recommended that addresses are passed in sorted in ascending order.
+///
+/// The process' ID should be provided in `pid`. To normalize addresses of the
+/// calling processes, `0` can be provided as a sentinel for the current
+/// process' ID.
+///
+/// Normalized addresses are reported in the exact same order in which the
+/// non-normalized ones were provided.
+pub fn normalize_user_addrs(addrs: Cow<'_, [Addr]>, pid: u32) -> Result<NormalizedUserAddrs> {
+    let mut addrs = addrs;
+
+    if !util::is_sorted(addrs.iter()) {
+        log::debug!("addresses to normalize not sorted; sorting in line...");
+        let () = addrs.to_mut().sort();
+    }
```

Comment:
I think one thing we can consider doing is having this check done inline with the main logic. That is, just make sure that we see addresses in ascending order. If that is not the case we sort and start over. That would mean we'd still allow for all inputs while not having to pay for an additional pass upfront.

Reply:
Updated.